Add `spring-data-sample` project (#19)
scalar-labs/scalardb-samples: spring-data-sample/README.md

# Spring Data integration Sample

This is a sample application for Spring Data integration with ScalarDB.

## Prerequisites
- Java (OpenJDK 8 or higher)
- Gradle
- Docker, Docker Compose

## Sample application

### Overview

> This article describes how to create an applicaiton for the same use case as [ScalarDB Sample](https://github.com/scalar-labs/scalardb-samples/tree/main/scalardb-sample) using Spring Data integration with ScalarDB.

Author: The main difference from `scalardb-sample`'s README are this `Overview` and command output examples (e.g. this application uses Jackson which doesn't put a space after comma)

The `sample.customers` table manages customers' information.
The `credit_limit` is the maximum amount of money a lender will allow each customer to spend using a credit card, and the `credit_total` is the amount of money that each customer has already spent by using the credit card.

The `sample.orders` table manages orders' information, and the `sample.statements` table manages the statements' information of the orders. Finally, the `sample.items` table manages items' information to be ordered.

The ER diagram for the schema is as follows:

![ERD](images/ERD.png)

### Transactions

The following five transactions are implemented in this sample application:

1. Getting customer information
3. Placing an order. An order is paid by a credit card. It first checks if the amount of the money of the order exceeds the credit limit. If the check passes, it records order histories and updates the `credit_total`
4. Getting order information by order ID
5. Getting order information by customer ID
6. Repayment. It reduces the amount of `credit_total`.

### Configuration

The configurations for the sample application is as follows:

```properties

> scalar.db.storage=cassandra

Author: Noticed this configuration should be replaced with Spring boot application's `application.properties`. I'll update this

Author: Done https://github.com/scalar-labs/scalardb-samples/pull/19/commits/ee4ea9e2ceb754b7cb6a19050d2c5e50da2ec7ff

> # Spring Data integration Sample

Reviewer: @josh-wong 
It would be great if you could take a look at the README from an English perspective.

> # Spring Data integration Sample

Reviewer 2: ```suggestion
# Spring Data Integration Sample
```

> This is a sample Spring Boot application for Spring Data integration with ScalarDB.

Reviewer 2: ```suggestion
This tutorial describes how to create a sample Spring Boot application by using Spring Data integration in ScalarDB.
```

> ## Prerequisites
> - Java (OpenJDK 8 or higher)

Reviewer 2: Many static site generators require a paragraph break after headings. I suggest adding one here.

```suggestion
## Prerequisites

- Java (OpenJDK 8 or higher)
```

- Gradle
- Docker, Docker Compose

## Sample application

### Overview

> This article describes how to create a Spring Boot applicaiton for the same use case as [ScalarDB Sample](https://github.com/scalar-labs/scalardb-samples/tree/main/scalardb-sample) using Spring Data integration with ScalarDB.

Reviewer 2: ```suggestion
This tutorial describes how to create a sample Spring Boot application for the same use case as [ScalarDB Sample](https://github.com/scalar-labs/scalardb-samples/tree/main/scalardb-sample) but by using Spring Data integration in ScalarDB.
```

> Please note that application-specific error handling, authentication processing, etc., are omitted in the sample application since it focuses on explaining how to use Spring Data integration.

Reviewer 2: ```suggestion
Please note that application-specific error handling, authentication processing, etc. are omitted in the sample application since this tutorial focuses on explaining how to use Spring Data integration.
```

> Please see [this document](https://scalar-labs.github.io/scalardb-sql/spring-data-guide.html) for the details.

Reviewer 2: ```suggestion
For details, please see [Guide of Spring Data integration with ScalarDB](https://scalar-labs.github.io/scalardb-sql/spring-data-guide.html).
```

### Schema

[The schema](schema.json) is as follows:

```json
{
  "sample.customers": {
    "transaction": true,
    "partition-key": [
      "customer_id"
    ],
    "columns": {
      "customer_id": "INT",
      "name": "TEXT",
      "credit_limit": "INT",
      "credit_total": "INT"
    }
  },
  "sample.orders": {
    "transaction": true,
    "partition-key": [
      "customer_id"
    ],
    "clustering-key": [
      "timestamp"
    ],
    "secondary-index": [
      "order_id"
    ],
    "columns": {
      "order_id": "TEXT",
      "customer_id": "INT",
      "timestamp": "BIGINT"
    }
  },
  "sample.statements": {
    "transaction": true,
    "partition-key": [
      "order_id"
    ],
    "clustering-key": [
      "item_id"
    ],
    "columns": {
      "order_id": "TEXT",
      "item_id": "INT",
      "count": "INT"
    }
  },
  "sample.items": {
    "transaction": true,
    "partition-key": [
      "item_id"
    ],
    "columns": {
      "item_id": "INT",
      "name": "TEXT",
      "price": "INT"
    }
  }
}
```

All the tables are created in the `sample` namespace.

> The `sample.customers` table manages customers' information.
> The `credit_limit` is the maximum amount of money a lender will allow each customer to spend using a credit card, and the `credit_total` is the amount of money that each customer has already spent by using the credit card.
> 
> The `sample.orders` table manages orders' information, and the `sample.statements` table manages the statements' information of the orders. Finally, the `sample.items` table manages items' information to be ordered.
> 

Reviewer 2: To help the reader skim the document, how about we make these descriptions into a bulleted listed?

```suggestion
- `sample.customers`: a table that manages customers' information
- `credit_limit`: the maximum amount of money a lender will allow each customer to spend when using a credit card
- `credit_total`: the amount of money that each customer has already spent by using the credit card
`sample.orders`: a table that manages order information
`sample.statements`: a table that manages order statement information
`sample.items`: a table that manages information of items to be ordered
```

> The ER diagram for the schema is as follows:

Reviewer 2: Does the acronym "ER" stand for "entity-relationship"? If so, how about we spell out the acronym instead?

![ERD](images/ERD.png)

### Transactions

The following five transactions are implemented in this sample application:

> 1. Getting customer information
> 3. Placing an order. An order is paid by a credit card. It first checks if the amount of the money of the order exceeds the credit limit. If the check passes, it records order histories and updates the `credit_total`
> 4. Getting order information by order ID
> 5. Getting order information by customer ID
> 6. Repayment. It reduces the amount of `credit_total`.

Reviewer 2: ```suggestion
1. Getting customer information
2. Placing an order by credit card (checks if the cost of the order is below the credit limit, then records order history and updates the `credit_total` if the check passes)
3. Getting order information by order ID
4. Getting order information by customer ID
5. Repayment (reduces the amount in the `credit_total`)
```

### Configuration

> The configurations for the sample Spring Boot application is as follows:

Reviewer 2: ```suggestion
Configurations for the sample Spring Boot application are as follows:
```

```application.properties
spring.datasource.driver-class-name=com.scalar.db.sql.jdbc.SqlJdbcDriver
spring.datasource.url=jdbc:scalardb:
?scalar.db.storage=cassandra\
&scalar.db.contact_points=localhost\
&scalar.db.username=cassandra\
&scalar.db.password=cassandra
```

> Since you use Cassandra in this sample application as mentioned above, you need to configure Cassandra settings in the configuration.

Reviewer 2: ```suggestion
Since this sample application uses Cassandra, as shown above, you need to configure your settings for Cassandra in this configuration.
```

> ## Set up

Reviewer 2: ```suggestion
## Setup
```

> You need to run the following `docker-compose` command:
> ```shell
> $ docker-compose up -d
> ```
> 
> This command starts Cassandra and loads the schema.
> Please note that you need to wait around more than one minute for the containers to be fully started.

Reviewer 2: ```suggestion
To start Cassandra and load the schema, you need to run the following `docker-compose` command:

```shell
$ docker-compose up -d
```

Please note that starting the containers may take more than one minute.
```

### Initial data

> You first need to load initial data with the following command:
> ```shell
> $ ./gradlew run --args="LoadInitialData"
> ```

Reviewer 2: ```suggestion
After the containers have loaded, you need to load the initial data by running the following command:

```shell
$ ./gradlew run --args="LoadInitialData"
```
```

> And the following data will be loaded:

Reviewer 2: ```suggestion
After the initial data has loaded, you will see the following:
```

- For the `sample.customers` table:

| customer_id | name          | credit_limit | credit_total |
|-------------|---------------|--------------|--------------|
| 1           | Yamada Taro   | 10000        | 0            |
| 2           | Yamada Hanako | 10000        | 0            |
| 3           | Suzuki Ichiro | 10000        | 0            |

- For the `sample.items` table:

| item_id | name   | price |
|---------|--------|-------|
| 1       | Apple  | 1000  |
| 2       | Orange | 2000  |
| 3       | Grape  | 2500  |
| 4       | Mango  | 5000  |
| 5       | Melon  | 3000  |

## Run the sample application

> Let's start with getting the customer information whose ID is `1`:
> ```shell

Reviewer 2: Let's add a paragraph break before the code block since static site generators may not render it properly without the break.

```suggestion
Let's start with getting information about the customer whose ID is `1`:

```shell
```

$ ./gradlew run --args="GetCustomerInfo 1"
...
{"customer_id":1,"name":"Yamada Taro","credit_limit":10000,"credit_total":0}
...
```

> Then, place an order for three apples and two oranges with customer ID `1`. Note that the format of order is `<Item ID>:<Count>,<Item ID>:<Count>,...`:
> ```shell

Reviewer 2: ```suggestion
Then, place an order for three apples and two oranges by using customer ID `1`. Note that the order format is `<Item ID>:<Count>,<Item ID>:<Count>,...`:

```shell
```

$ ./gradlew run --args="PlaceOrder 1 1:3,2:2"
...
{"order_id":"2358ab35-5819-4f8f-acb1-12e73d97d34e","customer_id":1,"timestamp":1677478005400}
...
```

> The command shows the order ID of the order.

Reviewer 2: ```suggestion
You can see that running this command shows the ID of the order.
```

> Let's check the details of the order with the order ID:
> ```shell

Reviewer 2: ```suggestion
Let's check the details of the order by using the order ID:

```shell
```

> $ ./gradlew run --args="GetOrder dea4964a-ff50-4ecf-9201-027981a1566e"

Reviewer 3: Should it be consistent with the order_id of the previous command?
```suggestion
$ ./gradlew run --args="GetOrder 2358ab35-5819-4f8f-acb1-12e73d97d34e"
```

...
{"order_id":"2358ab35-5819-4f8f-acb1-12e73d97d34e","timestamp":1677478005400,"customer_id":1,"customer_name":"Yamada Taro","statements":[{"item_id":5,"item_name":"Melon","price":3000,"count":1,"total":3000}],"total":3000}
...
```

> So, let's place an order again and get the order histories by customer ID `1`:
> ```shell

Reviewer 2: ```suggestion
Then, let's place another order and get the order history of customer ID `1`:

```shell
```

$ ./gradlew run --args="PlaceOrder 1 5:1"
...
{"order_id":"46062b16-b71b-46f9-a9ff-dc6b0991259b","customer_id":1,"timestamp":1677478201428}
...
$ ./gradlew run --args="GetOrders 1"
...

> [{"order_id":"2358ab35-5819-4f8f-acb1-12e73d97d34e","timestamp":1677478005400,"customer_id":1,"customer_name":"Yamada Taro","statements":[{"item_id":5,"item_name":"Melon","price":3000,"count":1,"total":3000}],"total":3000},{"order_id":"46062b16-b71b-46f9-a9ff-dc6b0991259b","timestamp":1677478201428,"customer_id":1,"customer_name":"Yamada Taro","statements":[{"item_id":5,"item_name":"Melon","price":3000,"count":1,"total":3000}],"total":3000}]

Reviewer 3: The first element of the array should describe the order of `PlaceOrder 1 1:3,2:2` I suppose?
```suggestion
[{"order_id":"2358ab35-5819-4f8f-acb1-12e73d97d34e","timestamp":1677478005400,"customer_id":1,"customer_name":"Yamada Taro","statements":[{"item_id":1,"item_name":"Apple","price":1000,"count":3,"total":3000},{"item_id":2,"item_name":"Orange","price":2000,"count":2,"total":4000}],"total":7000},{"order_id":"46062b16-b71b-46f9-a9ff-dc6b0991259b","timestamp":1677478201428,"customer_id":1,"customer_name":"Yamada Taro","statements":[{"item_id":5,"item_name":"Melon","price":3000,"count":1,"total":3000}],"total":3000}]
```

...
```

> These histories are ordered by timestamp in a descending manner.

Reviewer 2: ```suggestion
This order history is shown in descending order by timestamp.
```

> The current `credit_total` is `10000`, so it has reached the `credit_limit`.
> So, the customer can't place an order anymore due to the limit.

Reviewer 2: ```suggestion
The customer's current `credit_total` is `10000`. Since the customer has now reached their `credit_limit`, which was shown when retrieving their information, they cannot place anymore orders.
```

        at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at picocli.CommandLine.execute(CommandLine.java:2170)
        at sample.SampleApp.run(SampleApp.java:26)
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:768)
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:752)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:314)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1303)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1292)
        at sample.SampleApp.main(SampleApp.java:35)
...
```

> After repayment, the customer will be able to place an order again!

Reviewer 2: ```suggestion
After making a payment, the customer will be able to place orders again.
```

```shell
$ ./gradlew run --args="Repayment 1 8000"
...
$ ./gradlew run --args="GetCustomerInfo 1"
...
{"customer_id":1,"name":"Yamada Taro","credit_limit":10000,"credit_total":2000}
...
$ ./gradlew run --args="PlaceOrder 1 3:1,4:1"
...
{"order_id":"0350947a-9003-46f2-870e-6aa4b2df0f1f","customer_id":1,"timestamp":1677478728134}
...
```

> ## Clean up

Reviewer 2: ```suggestion
## Cleanup
```

> To stop Cassandra, run the following command:
> ```shell

Reviewer 2: ```suggestion
To stop Cassandra and shut down the sample application, run the following command:

```shell
```

3. Getting order information by order ID
4. Getting order information by customer ID
5. Repayment (reduces the amount in the `credit_total`)

### Configuration

Configurations for the sample Spring Boot application are as follows:

```application.properties
spring.datasource.driver-class-name=com.scalar.db.sql.jdbc.SqlJdbcDriver
spring.datasource.url=jdbc:scalardb:
?scalar.db.storage=cassandra\
&scalar.db.contact_points=localhost\
&scalar.db.username=cassandra\
&scalar.db.password=cassandra
```

Since this sample application uses Cassandra, as shown above, you need to configure your settings for Cassandra in this configuration.

## Setup

To start Cassandra and load the schema, you need to run the following `docker-compose` command:

```shell
$ docker-compose up -d
```

Please note that starting the containers may take more than one minute.

### Initial data

After the containers have loaded, you need to load the initial data by running the following command:

```shell
$ ./gradlew run --args="LoadInitialData"
```

> After the initial data has loaded, you will see the following:

Author: ```
- And the following data will be loaded:
+ After the initial data has loaded, you will see the following:
```

@josh-wong Noticed that users need to use Cassandra CLI or ScalarDB CLI tool to see the stored records and I think it's a bit hard. How about just describing the data has been loaded like as follows?

```
After the initial data has loaded, the following records should be stored in the tables:
```

Reviewer 2: @komamitsu Thank you for clarifying. LGTM!
